### PR TITLE
LA-26 - Students should be added/removed to the Opt-out LRS table on UI interaction

### PR DESCRIPTION
**Pull request by sandeepmjay**
The changes include:

Creation of backend api to process student opt-in/opt-out.
Ui-switch controller/factory updates for angular module.
Minor data model changes for opt-outs table.
 
JIRA ticket
https://jira.ets.berkeley.edu/jira/browse/LA-26

### CODE REVIEW

File: public/app/privacydashboard/datauses/datausesController.js [hunk 1]
```diff
@@ -43,6 +43,15 @@
           $scope.dataUses = dataUses.data;
         });
       };
+      /**
+       * Updates student's selection on data share opt-in/opt-out for specific projects.
+       * The datause object is with updated share selection is passed on to the api to process appropriately.
+       *
+       * @api private
+       */
+      $scope.toggleSwitch = function (dataUse){
```
Comment (johncrossman) [inline public/app/privacydashboard/datauses/datausesController.js]: @sandeepmjay stray space after `function`

File: public/app/privacydashboard/totalactivities/totalactivitiesController.js [hunk 2]
```diff
@@ -165,8 +165,9 @@
           // Convert the provided total activity numbers to a format readable by Highcharts
           $scope.totalActivitiesOptions.xAxis.categories = _.map($scope.totalActivities, 'period');
 
+          // TODO : Remove the ! condition for production. 
           // Change the current month to a dotted line
-          if ($scope.totalActivities[$scope.totalActivities.length - 1].current) {
+          if (!$scope.totalActivities[$scope.totalActivities.length - 1].current) {
```
Comment (johncrossman) [inline public/app/privacydashboard/totalactivities/totalactivitiesController.js]: @sandeepmjay can you say more on why we must "remove the ! condition for production"

File: public/app/privacydashboard/datauses/datausesController.js (hunk 1 above)
Comment (sandeepmjay) [inline public/app/privacydashboard/datauses/datausesController.js]: Fixed.

Comment (johncrossman) [inline public/app/privacydashboard/datauses/datausesController.js]: Ah, I guess it wants to live _after_ `(dataUser)`

File: public/app/privacydashboard/datauses/datausesFactory.js [hunk 3]
```diff
@@ -43,8 +43,20 @@
...
 
+    /**
+     * When a student makes changes on data share for specific projects create a post request to update the records suitably
+     *
+     * @return {Promise<Project[]>}                              $http promise returning the projects that potentially have access to the current user's data
+     */
+    var updateDataShare = function(dataUse) {
+
+     // enable this to post and update data share from the api
+     return $http.post('/api/user/datashare', dataUse);
```
Comment (johncrossman) [inline public/app/privacydashboard/datauses/datausesFactory.js]: @sandeepmjay extra space indent needed above

File: node_modules/lrs-core/lib/db.js [hunk 4]
```diff
@@ -264,10 +264,20 @@ var setUpModel = function(sequelize) {
...
-  User.belongsToMany(ReadCredential, {'through': 'opt_out'});
+  /**
+   * The `opt_out` table keeps track of students who have opted_out of the data use for Learning analytics projects
+   *
+   * @property  {String}      user_id                    The unique UUID of the learning activity statement
+   * @property  {String}      read_credential_id               The raw learning activity statement object
+   */
+   var Opt_out = module.exports.Opt_out = sequelize.define('opt_out',{
+
+   });
```
Comment (johncrossman) [inline node_modules/lrs-core/lib/db.js]: @sandeepmjay missing column definitions above?  Also, space needed after comma.

File: node_modules/lrs-users/lib/api.js [hunk 5]
```diff
@@ -336,3 +336,94 @@ var getDataUses = module.exports.getDataUses = function(ctx, callback) {
...
+* Validate and save user data share opt-in/opt-out request
+*
+* @param  {Object}           ctx                       Context containing the user for which to the project that have access to its data
+* @param  {Object}           dataUse                   dataUse details from which project id's and
+* @param  {Function}         callback                  Standard callback function
+* @param  {Object}           callback.err              An error that occurred, if any
+* @param  {Object[]}         callback.dataUses         The different projects that have access to the user's data and whether the user has opted out of this
+*/
+
+var saveUserOptOut = module.exports.saveUserOptOut = function (ctx, dataUse, callback){
```
Comment (johncrossman) [inline node_modules/lrs-users/lib/api.js]: @sandeepmjay stray space after `function`

File: node_modules/lrs-users/lib/api.js [hunk 6]
```diff
@@ -336,3 +336,94 @@ var getDataUses = module.exports.getDataUses = function(ctx, callback) {
...
+
+  if(dataUse.share === false){
+    // check if the opt-out details for the user and read credential already exists
+    getOrCreateOptOutDetails(ctx, dataUse, function (err, results){
```
Comment (johncrossman) [inline node_modules/lrs-users/lib/api.js]: @sandeepmjay the code style used in SuiteC would dictate: `function(err, results) {`.  Please let me know if you have different style in mind for `cloud-lrs`

File: node_modules/lrs-users/lib/api.js [hunk 7]
```diff
@@ -336,3 +336,94 @@ var getDataUses = module.exports.getDataUses = function(ctx, callback) {
...
+        return callback(err);
+      }
+      if(results){
+        log.info('The user %d has opt-out of data share for the project %d. Opt-out successful !', ctx.user.id, dataUse.id);
+        return callback();
+      }
+    });
+
+  }else if(dataUse.share === true){
+    // check if shre option is set to true. Remove the use from the opt-out table.
```
Comment (johncrossman) [inline node_modules/lrs-users/lib/api.js]: typo: shre

File: node_modules/lrs-users/lib/api.js [hunk 8]
```diff
@@ -336,3 +336,94 @@ var getDataUses = module.exports.getDataUses = function(ctx, callback) {
...
+    // This automatically enables data share option for the user on the selected project in the dataUse object
+    DB.Opt_out.destroy({
+      where: {
+        user_id: ctx.user.id,
+        read_credential_id: dataUse.id //this will be the ids that you want to delete
+      }
+    }).then(function(rowDeleted){ // rowDeleted will return number of rows deleted
+      if(rowDeleted === 1){
```
Comment (johncrossman) [inline node_modules/lrs-users/lib/api.js]: @sandeepmjay if we agree that code styles should be consistent then I'd scan this file for missing/stray spaces.  No biggie.

File: node_modules/lrs-users/lib/api.js [hunk 9]
```diff
@@ -336,3 +336,94 @@ var getDataUses = module.exports.getDataUses = function(ctx, callback) {
...
+    }, function(err){
+      log.error(err);
+    });
+    return callback();
+
+  }else{
+
+    // TODO : Handle deviations for wrong dataUse object in request body
+    // Log information if the dataUse object in request body is not correct.
+    log.info('The datause objects has a wrong share value. Share has to be a boolean.');
```
Comment (johncrossman) [inline node_modules/lrs-users/lib/api.js]: @sandeepmjay  grammar fix: "The datause objects have..."

File: public/app/privacydashboard/totalactivities/totalactivitiesController.js (hunk 2 above)
Comment (sandeepmjay) [inline public/app/privacydashboard/totalactivities/totalactivitiesController.js]: Added more context in comments

File: node_modules/lrs-users/lib/api.js [hunk 10]
```diff
@@ -336,3 +336,94 @@ var getDataUses = module.exports.getDataUses = function(ctx, callback) {
...
+
+  DB.Opt_out.findOrCreate(options).complete(function(err, data) {
+    if (err) {
+      log.error({'err': err}, 'Failed to get or create a user opt-out for data share');
+      return callback({'code': 500, 'msg': err.message});
+    }
+
+    var userOptOutDetails = data[0];
+    console.log('Creating/Retrieving Opt-out record in the table');
+    console.log(JSON.stringify(data[0]));
```
Comment (johncrossman) [inline node_modules/lrs-users/lib/api.js]: @sandeepmjay in the line above, might as well replace `data[0]` with `userOptOutDetails`

File: node_modules/lrs-users/lib/api.js (hunk 8 above)
Comment (sandeepmjay) [inline node_modules/lrs-users/lib/api.js]: Refactoring spaces and making it consistent with SuiteC. Good eye !

**Comment by johncrossman** [comment]
Thanks for talking me through certain bits, @sandeepmjay  Let me know when you've amended the PR per comments.

File: node_modules/lrs-users/lib/api.js (hunk 9 above)
Comment (sandeepmjay) [inline node_modules/lrs-users/lib/api.js]: Fixed

File: node_modules/lrs-users/lib/api.js (hunk 10 above)
Comment (sandeepmjay) [inline node_modules/lrs-users/lib/api.js]: Cleaned up logs used for debug purposes.

**Comment by sandeepmjay** [comment]
Code refactored to make spacing consistent. Comments updated. @johncrossman Thanks for the review. Go ahead with the merge at your convenience.

**Comment by johncrossman** [comment]
Thank you much, @sandeepmjay

**Pull request merged**